sklearn/datasets/openml.py

```diff
@@ -0,0 +1,415 @@
+    cached_get_data_description_by_id = mem(_get_data_description_by_id)
+    cached_get_data_features = mem(_get_data_features)
+    cached_download_data_arff = mem(_download_data_arff)
+
+    if not exists(data_home):
+        os.makedirs(data_home)
+
+    # check valid function arguments. data_id XOR (name, version) should be
+    # provided
+    if name is not None:
+        # OpenML is case-insensitive, but the caching mechanism is not
+        # convert all data names (str) to lower case
+        name = name.lower()
+        if data_id is not None:
+            raise ValueError(
+                "Dataset data_id={} and name={} passed, but you can only "
+                "specify a numeric data_id or a name, not "
+                "both.".format(data_id, name))
+        data_info = cached_get_data_info_by_name(name, version)
+        data_id = data_info['did']
+    elif data_id is not None:
+        # from the previous if statement, it is given that name is None
+        if version is not "active":
+            raise ValueError(
+                "Dataset data_id={} and version={} passed, but you can only "
+                "specify a numeric data_id or a version, not "
+                "both.".format(data_id, name))
+    else:
+        raise ValueError(
+            "Neither name nor data_id are provided. Please provide name or "
+            "data_id.")
+
+    data_description = cached_get_data_description_by_id(data_id)
+    if data_description['status'] != "active":
+        warn("Version {} of dataset {} is inactive, meaning that issues have"
+             " been found in the dataset. Try using a newer version.".format(
+                 data_description['version'], data_description['name']))
+
+    # download data features, meta-info about column types
+    features_list = cached_get_data_features(data_id)
+    features_dict = {feature['name']: feature for feature in features_list}
+
+    if target_column_name == "default-target":
+        target_column_name = _determine_default_target(features_list)
+
+    # TODO: stacking the content of the structured array
+    # this results in a copy. If the data was homogeneous
+    # and target at start or end, we could use a view instead.
+    data_columns = []
+    for feature in features_list:
+        # determine whether `feature` is a target
+        if (isinstance(target_column_name, string_types) and
+                feature['name'] == target_column_name):
+            is_target = True
+        elif (isinstance(target_column_name, list) and
+                feature['name'] in target_column_name):
+            is_target = True
+        else:
+            is_target = False
+
+        if ((not is_target) and feature['is_ignore'] == 'false' and
+                feature['is_row_identifier'] == 'false'):
+            data_columns.append(feature['name'])
+
+    arff_data = cached_download_data_arff(data_description['file_id'])['data']
```

Comment:
I think it's worth caching the numpy array version rather than ARFF. It makes more sense to have that load quickly. On the other hand, I'm not entirely sure how to deal with extracting the target if we do that, unless we keep it stored in a numpy struct array (is that what liac-arff returns? I've not yet checked)...

Reply:
Should we be cross checking the number of samples (if not other things) against the metadata's NumberOfInstances?

Reply:
> I think it's worth caching the numpy array version rather than ARFF. It makes more sense to have that load quickly. On the other hand, I'm not entirely sure how to deal with extracting the target if we do that, unless we keep it stored in a numpy struct array (is that what liac-arff returns? I've not yet checked)...

This is something that depends on whether or not we support sparse. 

> Should we be cross checking the number of samples (if not other things) against the metadata's NumberOfInstances?

I think not. First, we don't obtain the OpenML meta-data (I would not mind adding that feature though. ) Second, the mocked versions deviate highly from the original OpenML datasets as we removed many instances. Basically, except for the `NumberOfAttributes` I don't think any other meta-feature will yield the same result.

Reply:
Well, we *want* to support sparse. It's only a question of how easily it can make the first cut.

> Second, the mocked versions deviate highly from the original OpenML datasets as we removed many instances.

Which only means that the mocked versions need modification... The question is whether the fetcher should be doing data integrity checking. Maybe we should leave an issue open for it.